ingestion/src/metadata/ingestion/source/database/bigtable/metadata.py

```diff
@@ -0,0 +1,192 @@
+#  Copyright 2024 Collate
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#  http://www.apache.org/licenses/LICENSE-2.0
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+import traceback
+from typing import Dict, Iterable, List, Optional, Tuple, Union
+
+from google.cloud.bigtable import row_filters
+from google.cloud.bigtable.instance import Instance
+from google.cloud.bigtable.table import Table
+
+from metadata.generated.schema.entity.data.table import (
+    ConstraintType,
+    TableConstraint,
+    TableType,
+)
+from metadata.generated.schema.entity.services.connections.database.bigTableConnection import (
+    BigTableConnection,
+)
+from metadata.generated.schema.metadataIngestion.workflow import (
+    Source as WorkflowSource,
+)
+from metadata.ingestion.api.steps import InvalidSourceException
+from metadata.ingestion.ometa.ometa_api import OpenMetadata
+from metadata.ingestion.source.database.bigtable.client import MultiProjectClient
+from metadata.ingestion.source.database.bigtable.models import Row
+from metadata.ingestion.source.database.common_nosql_source import (
+    SAMPLE_SIZE as GLOBAL_SAMPLE_SIZE,
+)
+from metadata.ingestion.source.database.common_nosql_source import CommonNoSQLSource
+from metadata.ingestion.source.database.multi_db_source import MultiDBSource
+from metadata.utils.logger import ingestion_logger
+
+logger = ingestion_logger()
+
+MAX_COLUMN_FAMILIES = 100
+TABLE_SAMPLE_SIZE = 100
+
+
+class BigtableSource(CommonNoSQLSource, MultiDBSource):
+    """
+    Implements the necessary methods to extract database metadata from Google BigTable Source.
+    BigTable is a NoSQL database service for handling large amounts of data. Tha mapping is as follows:
+      project -> instance -> table -> column_family.column
+      (database) (schema)
+    For more infor about BigTable: https://cloud.google.com/bigtable/?hl=en
+    All data types are registered as bytes.
+    """
+
+    def __init__(self, config: WorkflowSource, metadata: OpenMetadata):
+        super().__init__(config, metadata)
+        self.client: MultiProjectClient = self.connection_obj
+
+        # ths instances and tables are cached to avoid making redundant requests to the API.
+        self.instances: Dict[str, Tuple[Instance]] = {}
+        self.tables: Dict[str, Dict[str, Tuple[Table]]] = {}
```

Comment:
nit - since this is quite a hard typing, would it make sense to have a custom pydantic model?

Reply:
I changed the types to be aliased so it is clear how the dictionaries are formed. If you still think a pydantic Model is better suited I am open to modifying but I need an example of how to model it either in the project or some draft.

Reply:
thanks, works with the aliases 👍